Content.Shared/Storage/SharedItemCounterSystem.cs

```diff
@@ -0,0 +1,59 @@
+﻿using System.Collections.Generic;
+using JetBrains.Annotations;
+using Robust.Shared.Containers;
+using Robust.Shared.GameObjects;
+
+namespace Content.Shared.Storage
+{
+    [UsedImplicitly]
+    public abstract class SharedItemCounterSystem : EntitySystem
+    {
+        /// <inheritdoc />
+        public override void Initialize()
+        {
+            base.Initialize();
+            SubscribeNetworkEvent<AfterStorageFillEvent>(OnStorageFill);
+            SubscribeLocalEvent<SharedItemCounterComponent, EntInsertedIntoContainerMessage>(HandleEntityInsert);
+            SubscribeLocalEvent<SharedItemCounterComponent, EntRemovedFromContainerMessage>(HandleEntityRemoved);
+        }
+
+        private void OnStorageFill(AfterStorageFillEvent args)
+        {
+            if (ComponentManager.HasComponent<SharedItemCounterComponent>(args.Entity.Uid)
+                && ComponentManager.TryGetComponent(args.Entity.Uid, out SharedAppearanceComponent appearance)
+                && args.ContainedEntities != null)
+            {
+                var allLayer = new List<EntityUid>();
+                foreach (var entity in args.ContainedEntities)
+                {
+                    allLayer.Add(entity.Uid);
+                }
+
+                appearance.SetData(StorageMapVisuals.AllLayers, new ListOfUids(allLayer));
+            }
+        }
+
+        private void HandleEntityRemoved(EntityUid uid, SharedItemCounterComponent _,
+            EntRemovedFromContainerMessage args)
+        {
+            UpdateSprite(args, false);
+        }
+
+        private void HandleEntityInsert(EntityUid uid, SharedItemCounterComponent _,
+            EntInsertedIntoContainerMessage args)
+        {
+            UpdateSprite(args, true);
+        }
+
+        private static void UpdateSprite(ContainerModifiedMessage args, bool show)
+        {
+            if (args.Container.Owner.TryGetComponent(out SharedAppearanceComponent? appearanceComponent))
+            {
+                appearanceComponent.SetData(StorageMapVisuals.LayerChanged, new ShowEntityData(args.Entity.Uid, show));
+                // Since visualizers can't be removed we just reset this key to empty value
+                // otherwise it would always check on each update
+                appearanceComponent.SetData(StorageMapVisuals.AllLayers, new ListOfUids(new List<EntityUid>()));
```

Comment:
hmmmmm this seems definitely not ideal/a little hacky, if you implemented PJB's idea with a queue of each container modified event that gets flushed on next tick you could get rid of the afterstoragefillevent and StorageMapVisuals.AllLayers entirely, but if you want to keep this then this call should go before the LayerChanged SetData call for sure. not sure if theres a better way to avoid this, maybe setting it to null (if null counts as an 'object' for the KV store??)